source/upyun/upyun.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/yunify/qscamel/record"
+)
+
+const upyunEndpoint = ".upaiyun.com"
+
+// SourceUpyun is upyun bucket type source
+type SourceUpyun struct {
+	BucketURL  string
+	bucketName string
+	up         *upyun.UpYun
+
+	lastModifiedTimes map[string]time.Time
+	objsChan          chan *upyun.FileInfo
+	marker            string
+}
+
+var errMissCredential = errors.New("Miss accessKeyID or secretAccessKey for upyun")
+
+// NewSourceUpyun creates an instance of SourceUpyun
+func NewSourceUpyun(bucketName, zone, accessKeyID, secretAccessKey string) (*SourceUpyun, error) {
+	if accessKeyID == "" || secretAccessKey == "" {
+		return &SourceUpyun{}, errMissCredential
+	}
+
+	up := upyun.NewUpYun(&upyun.UpYunConfig{
+		Bucket:   bucketName,
+		Operator: accessKeyID,
+		Password: secretAccessKey,
+	})
+
+	fmt.Println("http://" + bucketName + "." + zone + upyunEndpoint)
+
+	return &SourceUpyun{
+		BucketURL:         "http://" + bucketName + "." + zone + upyunEndpoint,
+		bucketName:        bucketName,
+		up:                up,
+		lastModifiedTimes: make(map[string]time.Time),
+		marker:            "",
+	}, nil
+}
+
+// GetSourceSites implements MigrateSource.GetSourceSites
+func (source *SourceUpyun) GetSourceSites(
+	threadNum int, logger *log.Logger, recorder *record.Recorder,
+) (sourceSites []string, objectNames []string, skipped []string, done bool, err error) {
+	sourceSites, objectNames, skipped = []string{}, []string{}, []string{}
+	err = nil
+
+	if source.marker == "" {
+		source.objsChan = make(chan *upyun.FileInfo, 10)
+		// upyun golang sdk does not suport to get part of all the objects to do advanced treatment.
+		// Inside the "List" interface, it Summarize all objects got from diffent http resp
+		// So we store all the objects only at first time.
+		// Then we use the sotred objects to do partial traversal regarding the number of concurrent tasks
+		getObjCfg := upyun.GetObjectsConfig{
+			Path:         "/",
+			ObjectsChan:  source.objsChan,
+			MaxListLevel: -1,
+		}
+		err = source.up.List(&getObjCfg)
+
+		if err != nil {
+			return sourceSites, objectNames, skipped, done, err
+		}
+
+		// If the implementation of the “List” interface is changed in the later version ，the marker will be a cursor.
+		// But at current version, it is just a flag which indicate that the information of objects has been acquired.
+		source.marker = "The information of objects has been acquired."
+	}
+
+	// If we got enough sourceSites as mouch as threadNum, return and begin to fetch these objects concurrently
+	for i := 0; i < threadNum; {
+
+		if len(source.objsChan) == 0 {
+			// If there is no more objects to fetch, the entire task is fished.
+			done = true
+			return sourceSites, objectNames, skipped, done, err
+		}
+
+		for object := range source.objsChan {
+
+			if object.IsDir {
+				logger.Infof(
+					"Skip dir object: %s", object.Name,
+				)
+				continue
+			}
+
+			sourceSite := strings.Join([]string{source.BucketURL, object.Name}, "/")
```

Comment:
source.Bucket is sth like “http://XXXXX”, use path.Join() will convert it into “http:/XXXXXX”, maybe we should use strings.Join to keep “//”.